sources/osgShader/Compiler.js

```diff
@@ -1288,15 +1305,32 @@ define( [
             return fragCol;
         },
 
+        createShadowCastFragmentShaderGraph: function () {
+            var frag = factory.getNode( 'glFragColor' );
+            factory.getNode( 'ShadowCast' ).setShadowCastAttribute( this._shadowCastAttribute ).inputs( {
+                exponent: this.getOrCreateUniform( 'float', 'exponent0' ),
+                exponent1: this.getOrCreateUniform( 'float', 'exponent1' ),
+                shadowDepthRange: this.getOrCreateUniform( 'vec4', 'Shadow_DepthRange' ),
+                fragEye: this.getOrCreateInputPosition()
+            } ).outputs( {
+                color: frag
+            } );
+            return frag;
+        },
 
         // this is the main function that will generate the
         // fragment shader. If you need to improve / add your own
         // you could inherit and override this function
         createFragmentShaderGraph: function () {
 
-
             var roots = [];
 
+            // depth cast
+            if ( this._isShadowCast ) {
```

Comment:
a lot of logic is inside the function instead would it be possible to use instead the StateSet ShaderGeneratorName to hook the compiler in the graph ?
This way we could inherit for example the shadow compiler and it would directly replace createFragmentShaderGraph ?
It's just an idea if we could avoid to overload the internal logic